File: hackerrank/problem solving/algorithms/2 - implementation/5 - breaking the records/breaking_the_records.ts
// https://www.hackerrank.com/challenges/breaking-best-and-worst-records/problem?isFullScreen=true

function main() {
  readLine();
  const [mostRecordBreaks, leastRecordBreaks]: number[] =
    breakingRecords(readNumbers());
  console.log(`${mostRecordBreaks} ${leastRecordBreaks}`);
}

function readNumbers(): number[] {
  return readLine()
    .split(' ')
    .map((x) => +x);
}

// n: length of scores
// T: O(n)
// S: O(1) extra space
function breakingRecords(scores: number[]): number[] {
  let mostPoints: number = scores[0];
  let leastPoints: number = scores[0];
  let mostRecordBreaks: number = 0;
  let leastRecordBreaks: number = 0;

  for (const score of scores)
    if (score > mostPoints) {
      mostPoints = score;
      mostRecordBreaks++;
    } else if (score < leastPoints) {
      leastPoints = score;
      leastRecordBreaks++;
    }

  return [mostRecordBreaks, leastRecordBreaks];
}

//////////////////////////////////////////////////

import readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });
let inputLines: string[] = [];
rl.on('line', (line: string) => inputLines.push(line));
rl.on('close', main);

function readLine(): string {
  return inputLines.shift()!;
}

File: hackerrank/problem solving/algorithms/2 - implementation/5 - breaking the records/breaking_the_records.py
# https://www.hackerrank.com/challenges/breaking-best-and-worst-records/problem?isFullScreen=true

def main():
    input()
    print(*breaking_records(read_numbers()))


def read_numbers():
    return list(map(int, input().split()))


def breaking_records(scores):
    most_points, least_points = scores[0], scores[0]
    most_record_breaks, least_record_breaks = 0, 0

    for score in scores[1:]:
        if score > most_points:
            most_points = score
            most_record_breaks += 1
        elif score < least_points:
            least_points = score
            least_record_breaks += 1

    return most_record_breaks, least_record_breaks


if __name__ == '__main__':
    main()
